perfkitbenchmarker/publisher.py

```diff
@@ -538,6 +548,87 @@ def PublishSamples(self, samples):
       vm_util.IssueRetryableCommand(copy_cmd)
 
 
+class ElasticsearchPublisher(SamplePublisher):
+  """Publish samples to an Elasticsearch server. Index and document type
+  will be created if they do not exist.
+
+  Attributes:
+    es_uri: String. i.e. http://IP:port. Default "http://localhost:9200"
+    es_index: String. Default "perfkit"
+    es_type: String. Default "result"
+  """
+  def __init__(self, es_uri=None, es_index=None, es_type=None):
+    self.es_uri = es_uri
+    self.es_index = es_index.lower()
+    self.es_type = es_type
+    self.mapping = {
+        "mappings": {
+            "result": {
+                "numeric_detection": True,
+                "properties": {
+                    "timestamp": {
+                        "type": "date",
+                        "format": "yyyy-MM-dd HH:mm:ss.SSSSSS"
+                    },
+                    "value": {
+                        "type": "double"
+                    }
+                },
+                "dynamic_templates": [{
+                    "strings": {
+                        "match_mapping_type": "string",
+                        "mapping": {
+                            "type": "string",
+                            "fields": {
+                                "raw": {
+                                    "type": "string",
+                                    "index": "not_analyzed"
+                                }
+                            }
+                        }
+                    }
+                }]
+            }
+        }
+    }
+
+  def PublishSamples(self, samples):
+    """Publish samples to Elasticsearch service"""
+    es = Elasticsearch([self.es_uri])
+    if not es.indices.exists(index=self.es_index):
+      es.indices.create(index=self.es_index, body=self.mapping)
+      logging.info('Create index %s and default mappings', self.es_index)
+    for sample in samples:
+      sample = sample.copy()
+      # Make timestamp understandable by ES and human.
+      sample['timestamp'] = self._HumanReadableTimestamp(sample['timestamp'])
+      # Keys cannot have dots for ES
+      sample = self._deDotKeys(sample)
```

Comment:
`self._deDotKeys(self, res)` modifies the dictionary you pass in rather than making a copy. Are we sure stuff happening after `PublishSamples` isn't going to rely on those original key names?